Prepare documentation for v0.1 release (#272)
tweag/topiary: README.md

# Topiary

![CI Status][ci-badge]

Topiary aims to be a uniform formatter for simple languages, as part of
the [Tree-sitter] ecosystem. It is named after the art of clipping or
trimming trees into fantastic shapes.

> ## Motivation

Reviewer: Before even this section we need a section explaining to the reader in which case he should care about Topiary. Something brief and to the point. There are two typical users: formatter authors, and formatter users. The value proposition is easier to articulate for authors, evidently. For users, we can explain the formatting philosophy (which we describe in more details below), and maybe the fact that it can conveniently serve as a formatter for many languages in their repository (ore repositories even).

Author: Added in fd18ebc1c1a

The style in which code is written has, historically, been mostly left
to personal choice. Of course, this is subjective by definition and has
led to many wasted hours reviewing formatting choices, rather than the
code itself. Prescribed style guides were an early solution to this,
spawning tools that lint a developer's formatting and ultimately leading
to automatic formatters. The latter were popularised by

> [`gofmt`][gofmt], whose developers had the insight that "good enough"

Reviewer: Maybe we ought to link to the slides which talk about design decisions somewhere.

For the record, I don't care too much about the absence of configurability (in fact Topiary is infinitely configurable: just clone the query file, and do the modifications that you want). It is important in Ormolu only because Haskell is so hard to parse and format that it's quite hard to test with configuration options (yes, changing from 2 to 4 spaces of indentation can conceivably break a Haskell output if not done carefully) but here, it doesn't matter too much.

Author: Added in fd18ebc1c1a

uniform formatting, imposed on a codebase, with no scope for subjective
adjustment, largely resolves these problems.

Topiary follows this trend by aspiring to be a "universal formatter
engine", which allows developers to not only automatically format their
codebases with a uniform style, but to define that style for new
languages using a [simple DSL][tree-sitter-query]. This allows for the
rapid prototyping of formatters, providing a [Tree-sitter
grammar][tree-sitter-parsers] is defined for that language.

## Design Principles

Topiary has been created with the following goals in mind:

* To target languages for which a [Tree-sitter
  grammar][tree-sitter-parsers] exist; whether officially supported by

>   the Tree-sitter project, or otherwise.

Reviewer: This paragraph sounds a bit cheesy. I guess we can say “use tree-sitter for parsing” if we want to make this point. Maybe explain that this avoids making yet another grammar for a formatter.

Author: Changed in fd18ebc1c1a

* To support a handful of languages, with "one true formatting style"
  for each, where no de facto formatter exists. These supported

>   formatting styles must:

Reviewer: As said above, “one true formatting style” is not a target. At any rate, they never exist in languages that have lived a while before their formatter was out. At any rate², “support[ing] a handful of languages” is not a design principle. It was a target for our MVP.

Author: Changed in fd18ebc1c1a

  * Be compatible with attested formatting styles used for that language
    in the wild.

>   * Be faithful to the author's intent, in terms of line spacing.

Reviewer: I don't know what this means.

Author: It's a paraphrasing of a point made in the `gofmt` slides:

> Respecting user intent is key.

i.e., If the author has written code that spans multiple lines, then they must have done so for a reason and that ought to be preserved. Hence Topiary's single-line vs. multi-line distinction.

Author: Reworded in fd18ebc1c1a

>   * Result in minimal diffs.

Reviewer: I know what this means. But I doubt a typical reader will.

Author: (This was copied from Ormolu's `README`, but I will expand...)

Author: Reworded in fd18ebc1c1a

  * Be idempotent. That is, formatting of already-formatted code doesn't
    change anything.

> * Code must be written in such a way that it's easy to modify and
>   maintain.
> 

Reviewer: Don't state the obvious, I feel it's insulting the readers' intelligence.

```suggestion
```

Author: (This was copied from Ormolu's `README`, but I will remove it.)

Author: Removed in fd18ebc1c1a

* Code and formatting styles must be well-tested and robust, so that the
  formatter can be used in large projects.

* For end users -- i.e., not formatting style developers -- the
  formatter should:

>   * Not permit any subjective configuration.

Reviewer: Again :slightly_smiling_face:

Author: IMO, the lack of configurability -- as an end user -- is important; its necessary for delivering on uniformity. Obviously the styles are customisable if you're so inclined, but I think that doing so pushes you into a different user category.

Author: Reworded in fd18ebc1c1a

  * Run efficiently.

  * Afford simple integration with other developer tools, such as
    editors and language servers.

## Language Support
<!-- Update this section as necessary on new developments/releases -->

[For now][topiary-issue4], the Tree-sitter grammars for the languages
that Topiary targets are statically linked. The formatting styles for
these languages come in two levels of maturity: supported (i.e., exposed
as a command line flag) and experimental (i.e., subject to change and/or
not yet production-ready).

#### Supported
* OCaml (both implementations and interfaces)
* JSON
* TOML

#### Experimental

* Rust

> * Bash

Reviewer: Be fair to your languages and give all of them a link.

Author: Added in fd18ebc1c1a

> * [Nickel][nickel]

Reviewer: You will be amused to know that 
```suggestion
* [Nickel]
```
works, because link names are case-insensitive in Markdown (I discovered recently).

Author: Done in fd18ebc1c1a

# Topiary

![CI Status][ci-badge]

Topiary aims to be a uniform formatter for simple languages, as part of
the [Tree-sitter] ecosystem. It is named after the art of clipping or
trimming trees into fantastic shapes.

Topiary is designed for formatter authors and formatter users. Authors
can create a formatter for a language without having to write their own

> parser. Users benefit from uniform code style and, potentially, the

Reviewer: Perhaps more importantly than the parser, is the whole engine as well.

I'm not sure the case for users is particularly good, but then again, we have little to say about that, I suppose. Maybe something along the line of “a single formatter for many languages with comparable style between languages”. After all, in our related work, we have a cornucopia of multi-language formatters. So it seems that's a value.

Author: Added in cd20b49a278a

convenience of using a single formatter tool across multiple languages
over their codebases.

## Motivation

The style in which code is written has, historically, been mostly left
to personal choice. Of course, this is subjective by definition and has
led to many wasted hours reviewing formatting choices, rather than the
code itself. Prescribed style guides were an early solution to this,
spawning tools that lint a developer's formatting and ultimately leading
to automatic formatters. The latter were popularised by

> [`gofmt`][gofmt], whose developers had [the insight][gofmt-slides] that
> "good enough" uniform formatting, imposed on a codebase, with no scope
> for subjective adjustment, largely resolves these problems.

Reviewer: ```suggestion
[`gofmt`][gofmt], whose developers had [the insight][gofmt-slides] that
"good enough" uniform formatting, imposed on a codebase largely resolves these problems.
```
I don't think that this bits adds relevant information.

Author: Removed in cd20b49a278a

Topiary follows this trend by aspiring to be a "universal formatter
engine", which allows developers to not only automatically format their
codebases with a uniform style, but to define that style for new
languages using a [simple DSL][tree-sitter-query]. This allows for the

> rapid prototyping of formatters, providing a [Tree-sitter

Reviewer: ```suggestion
fast development of formatters, providing a [Tree-sitter
```

Author: Changed in cd20b49a278a

grammar][tree-sitter-parsers] is defined for that language.

## Design Principles

Topiary has been created with the following goals in mind:

* Use [Tree-sitter] for parsing, to avoid writing yet another grammar
  for a formatter.

> * For bundled formatting styles -- for a handful of languages where no
>   de facto formatter exists -- to meet the following constraints:

Reviewer: ```suggestion
* For bundled formatting styles to meet the following constraints:
```

Author: Removed in cd20b49a278a

  * Be compatible with attested formatting styles used for that language
    in the wild.

>   * Be faithful to the author's intent: If code has been written such

Reviewer: ```suggestion
  * Be faithful to the author's intent: if code has been written such
```

Author: Changed in cd20b49a278a

    that it spans multiple lines, that decision is preserved.

  * Minimise changes between commits such that diffs focus mainly on the

>     code that's changed, rather that superficial artefacts.

Reviewer: I'm not sure that this is a good description of diff minimisation. The diff minimisation goal is to make a style where 1/ change on one line doesn't format others (I think that's what you are describing) 2/ (and, I'd say, it is more often where arbitrage is) a formatting style that doesn't require you to modify a line spuriously.

(2) is the reason for normalising terminators when we can (we currently can't in Ocaml, but this is why this style is desired):
```ocaml
[
  1;
  2;
]
```

Author: Expanded in cd20b49a278a

Options:

* `-l`, `--language <LANGUAGE>`\
  Which language to parse and format [possible values: `json`, `toml`,

>   `ocaml`, `ocaml-implementation`, `ocaml-interface`].

Reviewer: Is this part of the README generated?

Author: No. It probably should be... I'll add it as an issue.

Author: Done: Issue #277

  (#scope_id! "tuple")
)
```

> ...formats this piece of code:

Reviewer: I don't care either way, but this looks like a pretty unusual typographic style.

Author: I don't like the look of what appears to be a new sentence, but is actually just a follow-on clause interjected with a code block, starting with a lowercase letter. I add ellipses in this case as an indicator of this.

For example, the playground can be run in a tmux pane, with your editor
of choice open in another.

## Related Tools

### Tree-Sitter Specific

* [Syntax Tree Playground][tree-sitter-playground]: An interactive
  playground for experimenting with Tree-sitter and its query language.

### Meta and Multi-Language Formatters

* [format-all]: A formatter orchestrator for Emacs.
* [null-ls.nvim]: An LSP framework for Neovim that facilitates formatter
  orchestration.
* [prettier]: A formatter with support for multiple (web-development
  related) languages.
* [treefmt]: A general formatter orchestrator, which unifies formatters
  under a common interface.

### Related Formatters

> * [gofmt]: A (_the_) formatter for Go.

Reviewer: And major inspiration for the style of our formatters.

Author: Added in cd20b49a278a

* [ocamlformat]: A formatter for OCaml.
* [ocp-indent]: A tool to indent OCaml code.

> * [Ormolu]: A formatter for Haskell that follows similar design

Reviewer: ```suggestion
* [Ormolu]: Our formatter for Haskell that follows similar design
```

Author: Changed in cd20b49a278a

  principles as Topiary.

> * [rustfmt]: A formatter for Rust.

Reviewer: ```suggestion
* [rustfmt]: The _de facto_ standard formatter for Rust.
```

Author: Added in cd20b49a278a
